MusicXML importer: PartStaffs end up with too many (2) TimeSignatures in a non-first Measure

<!-- Version of music21:  print(music21.__version__) -->
**music21 version**

`7.1`   <!-- replace with actual version -->


**Problem summary**
<!-- Briefly: what undesired thing happens when what action is taken? -->
Attached is a MusicXML file that music21 wrote from a stream containing two PartStaffs (and a StaffGroup relating them).  It's an Organ score, with 5 measures, each of which has a new time signature: 7/8, 8/8, 7/8, 8/8, and 11/16, respectively.  I believe the MusicXML is correct, but I might be wrong.  It displays correctly in any case.  When I import this MusicXML file back into music21, however, while the first measure's time signature looks right, the second measure's time signature (and I'm guessing the others as well) shows up in the measure twice (in each PartStaff)

If I export a different MusicXML file from an otherwise identical Score containing Parts instead of PartStaffs, and then import _that_, all is well (I get one time signature at the beginning of each Measure in each Part).

I would note that neither MusicXML file appears to have the extra time signatures in it, so I'm guessing this is a bug in MusicXML import, that only happens when a part has more than one staff.

I have attached both MusicXML files for your enjoyment (zipped).

[LesCorps_Parts.musicxml.zip](https://github.com/cuthbertLab/music21/files/7173870/LesCorps_Parts.musicxml.zip)
[LesCorps_PartStaffs.musicxml.zip](https://github.com/cuthbertLab/music21/files/7173871/LesCorps_PartStaffs.musicxml.zip)

Thanks for the report. Reproducer:
```python
from music21 import converter, layout, meter, note, stream
ps1 = stream.PartStaff()
ps2 = stream.PartStaff()
sg = layout.StaffGroup([ps1, ps2])
s = stream.Score([ps1, ps2, sg])
for ps in ps1, ps2:
    ps.insert(0, meter.TimeSignature('3/1'))
    ps.repeatAppend(note.Note(type='whole'), 6)
    ps.makeNotation(inPlace=True)  # makes measures
    ps[stream.Measure][1].insert(meter.TimeSignature('4/1'))  # another time signature

fp = s.write()  # writes 4/1 time signatures in both hands
inn = converter.parse(fp, forceSource=True)
m2 = inn.parts.last()[stream.Measure][1]
assert len(m2[meter.TimeSignature]) == 1, m2.elements
```
```python
Traceback (most recent call last):
  File "/Users/jwalls/Desktop/repro.py", line 15, in <module>
    assert len(m2[meter.TimeSignature]) == 1, m2.elements
AssertionError: (<music21.meter.TimeSignature 4/1>, <music21.meter.TimeSignature 4/1>, <music21.note.Note C>, <music21.note.Note C>, <music21.note.Note C>, <music21.note.Rest whole>, <music21.bar.Barline type=final>)
```

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -2471,7 +2471,7 @@ class PartExporter(XMLExporterBase):
             self.midiChannelList = parent.midiChannelList  # shared list
             self.makeNotation = parent.makeNotation
 
-        self.higher_sibling_in_group: Optional[stream.PartStaff] = None
+        self.previous_sibling_in_group: Optional[stream.PartStaff] = None
 
         self.instrumentStream = None
         self.firstInstrumentObject = None
@@ -5660,7 +5660,7 @@ class MeasureExporter(XMLExporterBase):
 
         return mxAttributes
 
-    def _matches_earlier_sibling_in_group(self,
+    def _matches_previous_sibling_in_group(self,
                                           obj: base.Music21Object,
                                           attr='keySignature',
                                           comparison='__eq__') -> bool:
@@ -5672,7 +5672,7 @@ class MeasureExporter(XMLExporterBase):
         '''
         if self.parent is None:  # pragma: no cover
             return False
-        if self.parent.higher_sibling_in_group is None:
+        if self.parent.previous_sibling_in_group is None:
             return False
         # Not a foolproof measure lookup: see more robust measure
         # matching algorithm in PartStaffExporterMixin.processSubsequentPartStaff().
@@ -5680,7 +5680,7 @@ class MeasureExporter(XMLExporterBase):
         # see https://groups.google.com/g/music21list/c/ObNOanMQjJU/m/2LMPz5NAAwAJ
         if obj.measureNumber is None:  # pragma: no cover
             return False
-        maybe_measure = self.parent.higher_sibling_in_group.measure(obj.measureNumber)
+        maybe_measure = self.parent.previous_sibling_in_group.measure(obj.measureNumber)
         if maybe_measure is None:
             return False
         comparison_wrapper = getattr(obj, comparison)
@@ -6009,10 +6009,10 @@ class MeasureExporter(XMLExporterBase):
             mxDivisions.text = str(self.currentDivisions)
             self.parent.lastDivisions = self.currentDivisions
 
-        if m.keySignature is not None and not self._matches_earlier_sibling_in_group(
+        if m.keySignature is not None and not self._matches_previous_sibling_in_group(
                 m.keySignature, 'keySignature'):
             mxAttributes.append(self.keySignatureToXml(m.keySignature))
-        if m.timeSignature is not None and not self._matches_earlier_sibling_in_group(
+        if m.timeSignature is not None and not self._matches_previous_sibling_in_group(
                 m.timeSignature, 'timeSignature', comparison='ratioEqual'):
             mxAttributes.append(self.timeSignatureToXml(m.timeSignature))
         smts = list(m.getElementsByClass('SenzaMisuraTimeSignature'))

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -2447,7 +2447,9 @@ class PartExporter(XMLExporterBase):
     Object to convert one Part stream to a <part> tag on .parse()
     '''
 
-    def __init__(self, partObj: Union[stream.Part, stream.Score, None] = None, parent=None):
+    def __init__(self,
+                 partObj: Union[stream.Part, stream.Score, None] = None,
+                 parent: Optional[ScoreExporter] = None):
         super().__init__()
         # partObj can be a Score IF it has no parts within it.  But better to
         # always have it be a Part
@@ -2468,6 +2470,8 @@ class PartExporter(XMLExporterBase):
             self.refStreamOrTimeRange = parent.refStreamOrTimeRange
             self.midiChannelList = parent.midiChannelList  # shared list
             self.makeNotation = parent.makeNotation
+
+        self.higher_sibling_in_group: Optional[stream.PartStaff] = None
 
         self.instrumentStream = None
         self.firstInstrumentObject = None
@@ -2923,14 +2927,16 @@ class MeasureExporter(XMLExporterBase):
 
     ignoreOnParseClasses = {'LayoutBase', 'Barline'}
 
-    def __init__(self, measureObj=None, parent=None):
+    def __init__(self,
+                 measureObj: Optional[stream.Measure] = None,
+                 parent: Optional[PartExporter] = None):
         super().__init__()
-        if measureObj is not None:
-            self.stream = measureObj
-        else:  # no point, but...
+        if measureObj is None:  # no point, but...
             self.stream = stream.Measure()
+        else:
+            self.stream = measureObj
 
-        self.parent = parent  # PartExporter
+        self.parent = parent
         self.xmlRoot = Element('measure')
         self.currentDivisions = defaults.divisionsPerQuarter
 
@@ -5654,6 +5660,26 @@ class MeasureExporter(XMLExporterBase):
 
         return mxAttributes
 
+    def _matches_sibling(self,
+                         obj: base.Music21Object,
+                         attr='keySignature',
+                         comparison='__eq__') -> bool:
+        if self.parent is None:
+            return False
+        if self.parent.higher_sibling_in_group is None:
+            return False
+        # Not a foolproof measure lookup: see more robust measure
+        # matching algorithm in PartStaffExporterMixin.processSubsequentPartStaff().
+        # getElementsByOffset() would not be a perfect solution either,
+        # see https://groups.google.com/g/music21list/c/ObNOanMQjJU/m/2LMPz5NAAwAJ
+        if obj.measureNumber is None:
+            return False
+        maybe_measure = self.parent.higher_sibling_in_group.measure(obj.measureNumber)
+        if maybe_measure is None:
+            return False
+        comparison_wrapper = getattr(obj, comparison)
+        return comparison_wrapper(getattr(maybe_measure, attr))
+
     # -----------------------------
     # note helpers...
 
@@ -5977,20 +6003,21 @@ class MeasureExporter(XMLExporterBase):
             mxDivisions.text = str(self.currentDivisions)
             self.parent.lastDivisions = self.currentDivisions
 
-        if isinstance(m, stream.Measure):
-            if m.keySignature is not None:
-                mxAttributes.append(self.keySignatureToXml(m.keySignature))
-            if m.timeSignature is not None:
-                mxAttributes.append(self.timeSignatureToXml(m.timeSignature))
-            smts = list(m.getElementsByClass('SenzaMisuraTimeSignature'))
-            if smts:
-                mxAttributes.append(self.timeSignatureToXml(smts[0]))
+        if m.keySignature is not None and not self._matches_sibling(
+                m.keySignature, 'keySignature'):
+            mxAttributes.append(self.keySignatureToXml(m.keySignature))
+        if m.timeSignature is not None and not self._matches_sibling(
+                m.timeSignature, 'timeSignature', comparison='ratioEqual'):
+            mxAttributes.append(self.timeSignatureToXml(m.timeSignature))
+        smts = list(m.getElementsByClass('SenzaMisuraTimeSignature'))
+        if smts:
+            mxAttributes.append(self.timeSignatureToXml(smts[0]))
 
-            # For staves, see joinPartStaffs()
-            # TODO: part-symbol
-            # TODO: instruments
-            if m.clef is not None:
-                mxAttributes.append(self.clefToXml(m.clef))
+        # For staves, see joinPartStaffs()
+        # TODO: part-symbol
+        # TODO: instruments
+        if m.clef is not None:
+            mxAttributes.append(self.clefToXml(m.clef))
 
         found = m.getElementsByClass('StaffLayout')
         if found:

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -5660,10 +5660,16 @@ class MeasureExporter(XMLExporterBase):
 
         return mxAttributes
 
-    def _matches_sibling(self,
-                         obj: base.Music21Object,
-                         attr='keySignature',
-                         comparison='__eq__') -> bool:
+    def _matches_earlier_sibling_in_group(self,
+                                          obj: base.Music21Object,
+                                          attr='keySignature',
+                                          comparison='__eq__') -> bool:
+        '''
+        If this measure is part of a subsequent PartStaff in a joinable staff
+        group (e.g. the left hand of a keyboard part), then look up the
+        corresponding measure by number in the previous PartStaff, retrieve
+        the `attr` value there, and compare it to `obj` by `comparison`.
+        '''
         if self.parent is None:
             return False
         if self.parent.higher_sibling_in_group is None:
@@ -6003,10 +6009,10 @@ class MeasureExporter(XMLExporterBase):
             mxDivisions.text = str(self.currentDivisions)
             self.parent.lastDivisions = self.currentDivisions
 
-        if m.keySignature is not None and not self._matches_sibling(
+        if m.keySignature is not None and not self._matches_earlier_sibling_in_group(
                 m.keySignature, 'keySignature'):
             mxAttributes.append(self.keySignatureToXml(m.keySignature))
-        if m.timeSignature is not None and not self._matches_sibling(
+        if m.timeSignature is not None and not self._matches_earlier_sibling_in_group(
                 m.timeSignature, 'timeSignature', comparison='ratioEqual'):
             mxAttributes.append(self.timeSignatureToXml(m.timeSignature))
         smts = list(m.getElementsByClass('SenzaMisuraTimeSignature'))

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -2448,11 +2448,10 @@ class PartExporter(XMLExporterBase):
     '''
     _DOC_ATTR: Mapping[str, str] = {
         'previousPartStaffInGroup': '''
-        If the part being exported is a :class:`~music21.stream.base.PartStaff`,
-        this attribute will be used to store the immediately previous `PartStaff`
-        in the :class:`~music21.layout.StaffGroup`, if any. (E.g. if this is
-        the left hand, store a reference to the right hand.)
-        '''
+            If the part being exported is a :class:`~music21.stream.base.PartStaff`,
+            this attribute will be used to store the immediately previous `PartStaff`
+            in the :class:`~music21.layout.StaffGroup`, if any. (E.g. if this is
+            the left hand, store a reference to the right hand.)''',
     }
 
     def __init__(self,

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -23,7 +23,7 @@ import warnings
 from xml.etree.ElementTree import (
     Element, SubElement, ElementTree, Comment, fromstring as et_fromstring
 )
-from typing import Dict, List, Optional, Union
+from typing import Dict, List, Mapping, Optional, Union
 
 # external dependencies
 import webcolors
@@ -2446,6 +2446,14 @@ class PartExporter(XMLExporterBase):
     '''
     Object to convert one Part stream to a <part> tag on .parse()
     '''
+    _DOC_ATTR: Mapping[str, str] = {
+        'previousPartStaffInGroup': '''
+        If the part being exported is a :class:`~music21.stream.base.PartStaff`,
+        this attribute will be used to store the immediately previous `PartStaff`
+        in the :class:`~music21.layout.StaffGroup`, if any. (E.g. if this is
+        the left hand, store a reference to the right hand.)
+        '''
+    }
 
     def __init__(self,
                  partObj: Union[stream.Part, stream.Score, None] = None,
@@ -2471,7 +2479,7 @@ class PartExporter(XMLExporterBase):
             self.midiChannelList = parent.midiChannelList  # shared list
             self.makeNotation = parent.makeNotation
 
-        self.previous_sibling_in_group: Optional[stream.PartStaff] = None
+        self.previousPartStaffInGroup: Optional[stream.PartStaff] = None
 
         self.instrumentStream = None
         self.firstInstrumentObject = None
@@ -5660,7 +5668,7 @@ class MeasureExporter(XMLExporterBase):
 
         return mxAttributes
 
-    def _matches_previous_sibling_in_group(self,
+    def _matchesPreviousPartStaffInGroup(self,
                                           obj: base.Music21Object,
                                           attr='keySignature',
                                           comparison='__eq__') -> bool:
@@ -5672,7 +5680,7 @@ class MeasureExporter(XMLExporterBase):
         '''
         if self.parent is None:  # pragma: no cover
             return False
-        if self.parent.previous_sibling_in_group is None:
+        if self.parent.previousPartStaffInGroup is None:
             return False
         # Not a foolproof measure lookup: see more robust measure
         # matching algorithm in PartStaffExporterMixin.processSubsequentPartStaff().
@@ -5680,7 +5688,7 @@ class MeasureExporter(XMLExporterBase):
         # see https://groups.google.com/g/music21list/c/ObNOanMQjJU/m/2LMPz5NAAwAJ
         if obj.measureNumber is None:  # pragma: no cover
             return False
-        maybe_measure = self.parent.previous_sibling_in_group.measure(obj.measureNumber)
+        maybe_measure = self.parent.previousPartStaffInGroup.measure(obj.measureNumber)
         if maybe_measure is None:
             return False
         comparison_wrapper = getattr(obj, comparison)
@@ -6009,11 +6017,15 @@ class MeasureExporter(XMLExporterBase):
             mxDivisions.text = str(self.currentDivisions)
             self.parent.lastDivisions = self.currentDivisions
 
-        if m.keySignature is not None and not self._matches_previous_sibling_in_group(
-                m.keySignature, 'keySignature'):
+        if (m.keySignature is not None
+                and not self._matchesPreviousPartStaffInGroup(
+                    m.keySignature, 'keySignature'
+                )):
             mxAttributes.append(self.keySignatureToXml(m.keySignature))
-        if m.timeSignature is not None and not self._matches_previous_sibling_in_group(
-                m.timeSignature, 'timeSignature', comparison='ratioEqual'):
+        if (m.timeSignature is not None
+                and not self._matchesPreviousPartStaffInGroup(
+                    m.timeSignature, 'timeSignature', comparison='ratioEqual'
+                )):
             mxAttributes.append(self.timeSignatureToXml(m.timeSignature))
         smts = list(m.getElementsByClass('SenzaMisuraTimeSignature'))
         if smts:

--- a/music21/musicxml/m21ToXml.py
+++ b/music21/musicxml/m21ToXml.py
@@ -5670,7 +5670,7 @@ class MeasureExporter(XMLExporterBase):
         corresponding measure by number in the previous PartStaff, retrieve
         the `attr` value there, and compare it to `obj` by `comparison`.
         '''
-        if self.parent is None:
+        if self.parent is None:  # pragma: no cover
             return False
         if self.parent.higher_sibling_in_group is None:
             return False
@@ -5678,7 +5678,7 @@ class MeasureExporter(XMLExporterBase):
         # matching algorithm in PartStaffExporterMixin.processSubsequentPartStaff().
         # getElementsByOffset() would not be a perfect solution either,
         # see https://groups.google.com/g/music21list/c/ObNOanMQjJU/m/2LMPz5NAAwAJ
-        if obj.measureNumber is None:
+        if obj.measureNumber is None:  # pragma: no cover
             return False
         maybe_measure = self.parent.higher_sibling_in_group.measure(obj.measureNumber)
         if maybe_measure is None:

--- a/music21/musicxml/partStaffExporter.py
+++ b/music21/musicxml/partStaffExporter.py
@@ -1008,6 +1008,11 @@ class Test(unittest.TestCase):
         # Just two <attributes> tags, a 3/1 in measure 1 and a 4/1 in measure 2
         self.assertEqual(len(root.findall('part/measure/attributes/time')), 2)
 
+        # Edge cases -- no expectation of correctness, just don't crash
+        ps1[stream.Measure].last().number = 0  # was measure 2
+        root = self.getET(s)
+        self.assertEqual(len(root.findall('part/measure/attributes/time')), 3)
+
 
 if __name__ == '__main__':
     import music21

--- a/music21/musicxml/partStaffExporter.py
+++ b/music21/musicxml/partStaffExporter.py
@@ -254,6 +254,17 @@ class PartStaffExporterMixin:
                 MusicXMLWarning('Got overlapping StaffGroups; will not merge ANY groups.'))
             return []
 
+        # Finally, store a reference to earlier siblings (if any) on PartExporters
+        for group in deduplicatedGroups:
+            prior_part_staff = None
+            for part_staff in group:
+                for part_exporter in self.partExporterList:
+                    if part_exporter.stream is not part_staff:
+                        continue
+                    part_exporter.higher_sibling_in_group = prior_part_staff
+                    prior_part_staff = part_staff
+                    break
+
         return deduplicatedGroups
 
     def addStaffTagsMultiStaffParts(self, group: StaffGroup):
@@ -621,8 +632,9 @@ class PartStaffExporterMixin:
             partStaffRoot: Element = self.getRootForPartStaff(ps)
             # Remove PartStaff from export list
             # noinspection PyAttributeOutsideInit
-            self.partExporterList = [pex for pex in self.partExporterList
-                                        if pex.xmlRoot != partStaffRoot]
+            self.partExporterList: List[music21.musicxml.m21ToXml.PartExporter] = [
+                pex for pex in self.partExporterList if pex.xmlRoot != partStaffRoot
+            ]
 
     @staticmethod
     def moveMeasureContents(measure: Element, otherMeasure: Element, staffNumber: int):
@@ -976,6 +988,25 @@ class Test(unittest.TestCase):
         SX = ScoreExporter(b)
         SX.parse()
         SX.parse()
+
+    def testMeterChanges(self):
+        from music21 import layout
+        from music21 import meter
+        from music21 import note
+
+        ps1 = stream.PartStaff()
+        ps2 = stream.PartStaff()
+        sg = layout.StaffGroup([ps1, ps2])
+        s = stream.Score([ps1, ps2, sg])
+        for ps in ps1, ps2:
+            ps.insert(0, meter.TimeSignature('3/1'))
+            ps.repeatAppend(note.Note(type='whole'), 6)
+            ps.makeNotation(inPlace=True)  # makes measures
+            ps[stream.Measure][1].insert(meter.TimeSignature('4/1'))
+
+        root = self.getET(s)
+        # Just two <attributes> tags, a 3/1 in measure 1 and a 4/1 in measure 2
+        self.assertEqual(len(root.findall('part/measure/attributes/time')), 2)
 
 
 if __name__ == '__main__':

--- a/music21/musicxml/partStaffExporter.py
+++ b/music21/musicxml/partStaffExporter.py
@@ -145,9 +145,11 @@ class PartStaffExporterMixin:
     def joinableGroups(self) -> List[StaffGroup]:
         '''
         Returns a list of :class:`~music21.layout.StaffGroup` objects that
-        represent :class:`~music21.stream.PartStaff` objects that can be
+        represent :class:`~music21.stream.base.PartStaff` objects that can be
         joined into a single MusicXML `<part>`, so long as there exists a
-        `PartExporter` for it in `ScoreExporter.partExporterList`:
+        `PartExporter` for it in `ScoreExporter.partExporterList`.
+
+        Sets :attr:`~music21.musicxml.m21ToXml.PartExporter.previousPartStaffInGroup`.
 
         >>> s = stream.Score()
 
@@ -261,7 +263,7 @@ class PartStaffExporterMixin:
                 for part_exporter in self.partExporterList:
                     if part_exporter.stream is not part_staff:
                         continue
-                    part_exporter.previous_sibling_in_group = prior_part_staff
+                    part_exporter.previousPartStaffInGroup = prior_part_staff
                     prior_part_staff = part_staff
                     break
 

--- a/music21/musicxml/partStaffExporter.py
+++ b/music21/musicxml/partStaffExporter.py
@@ -261,7 +261,7 @@ class PartStaffExporterMixin:
                 for part_exporter in self.partExporterList:
                     if part_exporter.stream is not part_staff:
                         continue
-                    part_exporter.higher_sibling_in_group = prior_part_staff
+                    part_exporter.previous_sibling_in_group = prior_part_staff
                     prior_part_staff = part_staff
                     break
 

